File: 2449-maximum-number-of-robots-within-budget/maximum-number-of-robots-within-budget.go
func maximumRobots(chargeTimes []int, runningCosts []int, budget int64) int {
	var q []int
	result, s, left := 0, int64(0), 0
	for right, ct := range chargeTimes {
		for len(q) > 0 && ct > chargeTimes[q[len(q)-1]] {
			q = q[:len(q)-1]
		}

		q = append(q, right)
		s += int64(runningCosts[right])
		for len(q) > 0 && int64(chargeTimes[q[0]])+int64(right-left+1)*s > budget {
			s -= int64(runningCosts[left])
			for len(q) > 0 && q[0] <= left {
				q = q[1:]
			}
			left++
		}

		result = max(result, right-left+1)
	}
	return result
}

File: 2449-maximum-number-of-robots-within-budget/maximum-number-of-robots-within-budget.rs
use std::collections::vec_deque;

impl Solution {
    pub fn maximum_robots(charge_times: Vec<i32>, running_costs: Vec<i32>, budget: i64) -> i32 {
        let mut q = vec_deque::VecDeque::new();
        let mut result = 0;
        let mut s: i64 = 0;
        let mut left = 0;

        for (right, &ct) in charge_times.iter().enumerate() {
            while !q.is_empty() && ct > charge_times[*q.back().unwrap()] {
                q.pop_back();
            }

            q.push_back(right);
            s += running_costs[right] as i64;

            while !q.is_empty() && charge_times[q[0]] as i64 + (right - left + 1) as i64 * s as i64 > budget {
                s -= running_costs[left] as i64;

                while !q.is_empty() && q[0] <= left {
                    q.pop_front();
                }

                left += 1;
            }

            result = result.max(right - left + 1);
        }

        result as i32
    }
}